MarshalJSONWithOptions

pboyer opened the pull request:
Fixes #5.

The implementation strategy largely follows the approach suggested in #8. The only exception is that I added a new method `allKnownFieldTags` instead of passing a flag to `knownFieldTags`. This is mainly because I think bool flags hurt readability at the callsite (e.g. `foo(true, false)`), but I'm happy to change it if you'd like.

How would you like this to be tested? It's obviously a bit thin at the moment.

This requires #17 to be merged first, otherwise the diff is largely unreadable.

jhump commented on dynamic/json.go:
```diff
@@ -56,18 +59,34 @@ func (m *Message) marshalJSON(b *indentBuffer) error {
 
+		v, ok := m.values[itag]
+		if !ok {
+			if emitDefaults {
```
I don't think you need to double-check this flag: we should only ever get here if `emitDefaults`. (Also means you can remove the `emitDefaults` local variable and inline into its one other use.)

It may be possible to incorrectly reach this line through some concurrency bug in calling code, but I don' think we want to go out of our way to handle that (and, if we did, the right way to address it would probably be to `continue` if `!emitDefaults`...)

pboyer replied:
👍

pboyer commented on dynamic/json.go:
```diff
@@ -27,26 +27,29 @@ type MarshalJSONOpts struct {
 }
 
-func (m *Message) MarshalJSON() ([]byte, error) {
+func (m *Message) MarshalJSONCustom(opts MarshalJSONOpts) ([]byte, error) {
```
You got it.

pboyer commented:
@jhump I've addressed your comments and made two additional changes.

1. I was misusing `*indentBuffer.writeString` to write `null`. I didn't notice it writes a proper, escaped string when I want to write that string in raw form.
2. I changed `MarshalJSONOpts` to `MarshalJSONOptions` for consistency.

pboyer commented:
LOL. Thanks for the quick merge! I didn't even get the chance to finish my note. Cheers and happy Friday! :D
